command/agent/retry_join_test.go

```diff
@@ -166,10 +185,48 @@ func TestRetryJoin_Server_MixedProvider(t *testing.T) {
 	joiner.RetryJoin(serverJoin)
 
 	require.Equal(2, len(output))
-	require.Equal("provider=aws, tag_value=foo", mockDiscover.ReceivedAddrs)
+	require.Equal("provider=aws, tag_value=foo", mockDiscover.ReceivedConfig)
 	require.Equal(stubAddress, output[0])
 }
 
+func TestRetryJoin_AutoDiscover(t *testing.T) {
+	ci.Parallel(t)
+	require := require.New(t)
+
+	var output []string
+	mockJoin := func(s []string) (int, error) {
+		output = s
+		return 0, nil
+	}
+
+	// 'exec=*'' tests go-netaddr
+	// 'localhost' also tests get-netaddr by ensuring it resolves to "::1" and "127.0.0.1"
+	// '100.100.100.100' ensures that bare IPs are used as-is
+	serverJoin := &ServerJoin{
+		RetryMaxAttempts: 1,
+		RetryJoin:        []string{"exec=echo 127.0.0.1", "dns=localhost", "100.100.100.100", "provider=aws, tag_value=foo"},
+	}
+
+	mockDiscover := &MockDiscover{}
+	mockNetaddrs := &MockNetaddrs{}
+	joiner := retryJoiner{
+		autoDiscover:  autoDiscover{goDiscover: mockDiscover, netAddrs: mockNetaddrs},
+		serverJoin:    mockJoin,
+		serverEnabled: true,
+		logger:        testlog.HCLogger(t),
+		errCh:         make(chan struct{}),
+	}
+
+	joiner.RetryJoin(serverJoin)
+
+	require.Equal(4, len(output)) // [127.0.0.1 127.0.0.1 100.100.100.100 127.0.0.1]
+	require.Equal("100.100.100.100", output[2])
+	require.Equal(2, len(mockNetaddrs.ReceivedConfig))
+	require.Equal("exec=echo 127.0.0.1", mockNetaddrs.ReceivedConfig[0])
+	require.Equal("localhost", mockNetaddrs.ReceivedConfig[1])
+	require.Equal("provider=aws, tag_value=foo", mockDiscover.ReceivedConfig)
```

Comment:
We haven't migrated all the tests yet but for all new tests we're using [`shoenig/test`](https://pkg.go.dev/github.com/shoenig/test).

```suggestion
	must.Len(t, 4, output) // [127.0.0.1 127.0.0.1 100.100.100.100 127.0.0.1]
	must.Eq(t, "100.100.100.100", output[2])
	must.Len(t, 2, mockNetaddrs.ReceivedConfig)
	must.Eq(t, "exec=echo 127.0.0.1", mockNetaddrs.ReceivedConfig[0])
	must.Eq(t, "localhost", mockNetaddrs.ReceivedConfig[1])
	must.Eq(t, "provider=aws, tag_value=foo", mockDiscover.ReceivedConfig)
```

(Same for those above.)

Also, given that the order is guaranteed we could probably combine the first two assertions into:

```go
	must.Eq(t, []string{"127.0.0.1", "127.0.0.1", "100.100.100.100", "127.0.0.1"}, output)
```

Reply:
(This comment accidentally got marked resolved but wasn't actually fixed)